R/use_crate.R

```diff
@@ -0,0 +1,169 @@
+#' Add dependencies to a Cargo.toml manifest file
+#'
+#' Analogous to `usethis::use_package()` but for crate dependencies.
+#'
+#' @param crate character scalar, the name of the crate to add
+#' @param features character vector, a list of features to include from the
+#' crate
+#' @param git character scalar, the full URL of the remote Git repository
+#' @param version character scalar, the version of the crate to add
+#' @param optional boolean scalar, whether to mark the dependency as optional 
+#' (FALSE by default)
+#' @param path character scalar, the package directory
+#'
+#' @details
+#' For more details regarding these and other options, see the 
+#' \href{https://doc.rust-lang.org/cargo/commands/cargo-add.html}{Cargo docs}
+#' for `cargo-add`.
+#'
+#' @return `NULL`, invisibly
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # add to [dependencies]
+#' use_crate("serde")
+#'
+#' # add to [dependencies] and [features]
+#' use_crate("serde", features = "derive")
+#'
+#' # add to [dependencies] using github repository as source
+#' use_crate("serde", git = "https://github.com/serde-rs/serde")
+#'
+#' # add to [dependencies] with specific version
+#' use_crate("serde", version = "1.0.1")
+#' 
+#' # add to [dependencies] with optional compilation
+#' use_crate("serde", optional = TRUE)
+#' }
+use_crate <- function(
+    crate,
+    features = NULL,
+    git = NULL,
+    version = NULL,
+    optional = FALSE,
+    path = "."
+){
+
+  # check crate
+  if (!rlang::is_character(crate) && length(crate) != 1){
+
+    info <- paste(
+      "You supplied an object of class {class(crate)[1]}",
+      "with length {length(crate)}."
+    )
+
+    cli::cli_abort(
+      "{.var crate} should be a length one character vector.",
+      "i" =  info,
+      class = "rextendr_error"
+    )
+
+  }
+
+  # check features
+  if (!is.null(features) && !rlang::is_character(features)){
+
+    cli::cli_abort(
+      "{.var features} should be a character vector.",
+      "i" = "You supplied an object of class {class(features)[1]}.",
+      class = "rextendr_error"
+    )
+
+  }
+
+  # check git
+  if (!is.null(git)){
+
+    if (!rlang::is_character(git) && length(git) != 1){
+
+      info <- paste(
+        "You supplied an object of class {class(git)[1]}",
+        "with length {length(git)}."  
+      )
+
+      cli::cli_abort(
+        "{.var git} should be a length one character vector.",
+        "i" = info,
+        class = "rextendr_error"
+      )
+
+    }
+
+  }
+
+  # check version
+  if (!is.null(version)){
```

Comment:
Yeah, good point! But it raises the question: Is this validation logic something you'd also like to see for the whole package?

Reply:
I recommend using the standalone rlang checks. I am using these in all of my packages now. They only rely on R lang and make everything much much better to work with.

In the below I modified the existing standalones to use `rlang::` instead of the default which requires `#' @import rlang` 
https://github.com/R-ArcGIS/arcgisgeocode/blob/main/R/import-standalone-types-check.R
https://github.com/R-ArcGIS/arcgisgeocode/blob/main/R/import-standalone-obj-type.R

Reply:
I'm happy to implement Josiah's suggestion, but I would need the go-ahead from one of you folks in charge. And there's also the question if we would want to implement these validation tools everywhere in the package.

Reply:
You can put them in a separate file and start using here. We can slowly adopt em where applicable (out of scope of this PR).